tools/flaky_test_monitor/level3/process_summary3_results.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"encoding/json"
+	"flaky-test-monitor/common"
+	"os"
+	"sort"
+)
+
+// processSummary3TestRun processes a level 2 summary and produces level 3 summary which summarizes:
+// most failed tests, tests with no-results, longest running tests.
+func processSummary3TestRun(level2FilePath string, propertyFileDirectory string) common.TestSummary3 {
+
+	config := common.ReadProperties(propertyFileDirectory)
+
+	var testSummary2 common.TestSummary2
+
+	level2JsonBytes, err := os.ReadFile(level2FilePath)
+	common.AssertNoError(err, "error reading level 2 json")
+
+	err = json.Unmarshal(level2JsonBytes, &testSummary2)
+	common.AssertNoError(err, "error unmarshalling level 2 test run")
+
+	// there should be at least 1 level 2 test result in the supplied file
+	// if the json format is different in the supplied file, there won't be a marshalling error thrown
+	// this is an indirect way to tell if the json format was wrong (i.e. not a level 2 json format)
+	if len(testSummary2.TestResults) == 0 {
+		panic("invalid summary 2 file - no test results found")
+	}
+
+	// create lists to keep track of 3 main things
+	// 1. tests with no-results (ordered by most no-results)
+	// 2. tests with failures (ordered by most failures)
+	// 3. tests with durations > 0 (ordered by longest durations)
+
+	noResultsTRS := []common.TestResultSummary{}
+	failuresTRS := []common.TestResultSummary{}
+	durationTRS := []common.TestResultSummary{}
+
+	// go through all level 2 test results to figure out grouping for tests with
+	// most failures, no-results, longest running
+	for _, trs := range testSummary2.TestResults {
+		if trs.NoResult > 0 {
+			noResultsTRS = insertSortedByNoResult(noResultsTRS, *trs)
+		}
+		if trs.Failed > 0 && trs.FailureRate >= config.FailureThresholdPercent {
+			failuresTRS = insertSortedByFailureRate(failuresTRS, *trs)
+		}
+		if trs.AverageDuration > 0 && trs.AverageDuration >= config.DurationThresholdSeconds {
+			durationTRS = insertSortedByDuration(durationTRS, *trs)
+		}
+	}
+
+	var testSummary3 common.TestSummary3
+	testSummary3.NoResults = noResultsTRS
+
+	// total # of failed tests that satisfy min failure threshold
+	testSummary3.MostFailuresTotal = len(failuresTRS)
+
+	// check if # of failures exceeded max failures to return
+	if len(failuresTRS) > config.FailuresSliceMax {
+		// truncate slice to return only the first config.FailuresSliceMax failures
+		failuresTRS = failuresTRS[:config.FailuresSliceMax]
+	}
+	testSummary3.MostFailures = failuresTRS
+
+	// total # of long tests that satisfy min duration threshold
+	testSummary3.LongestRunningTotal = len(durationTRS)
+
+	// check if # of durations exceeded max durations to return
+	if len(durationTRS) > config.DurationSliceMax {
+		// truncate slice to return only the first config.DurationSliceMax durations
+		durationTRS = durationTRS[:config.DurationSliceMax]
+	}
+	testSummary3.LongestRunning = durationTRS
```

Comment:
I feel like we can make this logic way simpler with https://pkg.go.dev/sort#Slice.

We can just sort the level 2 results multiple different ways, and output the JSON each time to a separate file. This also makes it a lot easier to add different sorting criteria imo.

Reply:
And we won't need any of these helper functions below

Reply:
Agreed. Updated to use `sort.Slice()`

https://github.com/onflow/flow-go/pull/1793/commits/45f2adf7f0396d0624d35112d28b5598179eb0b1